HDF5 standards for input data and parameter values

dilpath opened the pull request:
Descriptions of the two new standards are provided in "format.md".

Tagging here since I can't select you as a reviewer... @FFroehlich 

- changed the naming for the `MLModel` standard from `PetabScimlStandard` to `MLModelStandard`
- all schemas are now provided in the JSON schema format instead. No strong preference, just for consistency

There are also `input_id`s in the input data HDF5 files, which I designed to be specified directly into the MLModel YAML as opposed to the mapping table that we previously agreed on, but this could be discussed in #17 .

sebapersson commented on docs/src/assets/parameter_values_schema.json:
```diff
@@ -0,0 +1,60 @@
+{
+    "$defs": {
+        "SingleLayerParameterValues": {
+            "description": "The values for the parameters in a single model layer.",
```
```suggestion
            "description": "The values for the parameters in a single layer.",
```

sebapersson commented on docs/src/assets/mlmodel_schema.json:
```diff
@@ -0,0 +1,149 @@
+{
+    "$defs": {
+        "Input": {
+            "description": "Specify (transformations of) the input layer.",
```
What do you mean with transformation of the input layer here?

sebapersson replied:
I think we should drop this, as I think transforming the data into appropriate format is outside the scope.

sebapersson commented on docs/src/assets/mlmodel_schema.json:
```diff
@@ -0,0 +1,149 @@
+            "type": "object"
+        }
+    },
+    "description": "An easy-to-use format to specify simple deep ML models.\n\nThere is a function to export this to a PyTorch module, or to YAML.",
```
```suggestion
    "description": "An easy-to-use format to specify simple deep ML models.\n\nThere is a function to export this to a PyTorch module, or to YAML file.",
```

sebapersson commented on docs/src/assets/parameter_values_schema.json:
```diff
@@ -0,0 +1,60 @@
+            "type": "object"
+        }
+    },
+    "description": "Parameter values for an ML model.",
```
I think we generally do not use ML model in the spec, maybe better to use neural network model as often as possible? (as ML can imply other things, e.g. Random forest)

dilpath replied:
Yes you're right about the spec. It is mostly implied there that the ML models are neural networks.

However, until now, we haven't explicitly written that the models expressed in Lux.jl or Equinox scripts are necessarily neural networks. I don't see a need to make this restriction, and therefore the spec could refer to ML models more broadly. So then, these parameter files would also be for ML models, not just ANNs.

But also fine for me if we decide to imply (or explicitly write) in the spec that the ML part of our SciML models are always neural networks.

sebapersson replied:
I think we should consistently keep it at neural networks.

sebapersson commented on docs/src/format.md:
```diff
@@ -36,20 +36,26 @@ A neural network model must consist of two parts to be compatible with the PEtab
+### [Neural Network Parameter Values](@id hdf5_ps_structure)
 
-All parameters for a neural network model are stored in an HDF5 file, with the file path specified in the problem [YAML file](@ref YAML_file). The HDF5 parameter file is expected to have the following structure for an arbitrary number of layers:
+Parameter values of frozen or pre-trained layers, or post-calibration parameter values, are stored in the HDF5 format, and included in the problem via the problem [YAML file](@ref YAML_file). The HDF5 file should contain a list of entries, where each entry specifies the parameter values for a single layer. Each parameter value includes the name of the parameter as well, e.g. `weight` for the weig ...
```
```suggestion
Parameter values for frozen or pre‑trained layers and post‑calibration parameters are stored in HDF5 format and and included in the problem via the [YAML file](@ref YAML_file). The HDF5 file must contain a list of entries, each representing a single layer. For each layer, the parameter array identifier(s) (e.g., `weight` and optionally `bias` for a PyTorch `Linear` layer) and their values must be provided.
```

dilpath replied:
Should
> (e.g., `weight` and optionally `bias` for a PyTorch `Linear` layer)

instead be
> (e.g., `weight` and/or `bias` for a PyTorch `Linear` layer)

?

i.e. it's fine to only supply one of these, right?

sebapersson replied:
No, because weight is mandatory for a Linear layer, while bias is optional.

dilpath replied:
Yes, but the weight can be trained? e.g. maybe the user only wants to transfer the bias from a previous run, but re-train the weight. Or do we exclude this possibility?

sebapersson replied:
Ahh good point, but this is excluded, as currently in the parameter table section it is stated that freezing a layer is not allowed.

dilpath replied:
Let's say I have a foundation model that extracts information from images of cells (e.g. width, circumference, roundness, neighbors, ...). And now I want to use it to specify the initial condition of an ODE system.

There will be the foundation model that extracts information, and the contextualizing model that generates the mechanistic-model--specific initial condition.
Do I need to keep these components separated as distinct models, in order to use the frozen foundation model with different mechanistic models? i.e.
```
image -> foundation model (frozen) -> contextualizing model (trained) -> ODE initial condition
```

Fine for me but maybe we could provide an example in the docs, so people can see that frozen layers are possible, if the user separates them into separate ML models.

---

But if this is possible, why not simply allow frozen layers or parameters?

The docs do state the freezing a layer is not possible, but it looks like a contradiction. The docs also state
> When parsing, deeper nested levels (e.g., netId.layerId) have precedence for nominal values, priors, etc.

If we have this idea of "deeper nested levels", why doesn't it apply to `estimate` too? Seems weird to me that I can specify layer-specific priors, but not layer-specific freezing.

sebapersson replied:
This will be a bit of a pain to implement, but you have a valid point. Frozen layers, and even frozen arrays, is something we should allow. Feel free to remove that it is now allowed in the spec, and I will put it on my TODO to add appropriate test-cases.

sebapersson commented on docs/src/format.md:
```diff
@@ -36,20 +36,26 @@ A neural network model must consist of two parts to be compatible with the PEtab
 
+The schema is provided as [JSON schema](assets/parameter_data_schema.json). Currently, validation is only provided via the PEtab SciML library.
+
 The indexing convention and naming for `arrayId` depend on the neural network model library:
```
```suggestion
The indexing convention and naming for `framework_parameter_name` depend on the neural network model library:
```

sebapersson commented on docs/src/format.md:
```diff
@@ -58,16 +64,39 @@ The indexing convention and naming for `arrayId` depend on the neural network mo
+### [Neural Network Input Data](@id hdf5_input_structure)
 
-Potential neural network input files are expected to have the following format:
+Input data for ML models are specified in the HDF5 format. The HDF5 file should contain a list of entries, where each entry specifies an input ID and the datasets for that input ID. Each dataset is the data itself, and optionally the experiment IDs that the data applies to. If no experiment IDs are specified, then the dataset will be utilized in all experiments. A single input cannot be assigned  ...
```
```suggestion
Array input data for neural network models is specified in HDF5 format. Each HDF5 file should contain a list of entries, with each entry defining an input Id and its associated datasets. Each dataset consists of the data array and, optionally, the experiment Ids to which it applies. If no experiment Ids are provided, the dataset will be applied to all experiments. Multiple datasets may not be assigned to the same input for a single experimental condition.
```

sebapersson commented on docs/src/format.md:
```diff
@@ -76,11 +105,9 @@ As with [parameters](@ref hdf5_ps_structure), the indexing depends on the neural
 ### [YAML Network file format](@id YAML_net_format)
 
-The `petab_sciml` library provides a YAML neural network file format for model exchange. The YAML format follows PyTorch conventions for layer names and arguments. The schema is provided as [YAML-formatted JSON schema](assets/net_schema.yaml), which enables easy validation with various third-party tools.
+The `petab_sciml` library provides a YAML neural network file format for model exchange. The YAML format follows PyTorch conventions for layer names and arguments. The schema is provided as [JSON schema](assets/mlmodel_schema.json), which enables easy validation with various third-party tools.
```
```suggestion
The `petab_sciml` library provides a YAML neural network file format for model exchange. The YAML format follows PyTorch conventions for layer names and arguments. The schema is provided as [JSON schema](assets/mlmodel_schema.json), which enables validation with various third-party tools.
```

sebapersson reviewed (changes requested):
Thanks a lot! Left a few minor comments.

I think the input format looks good, and once this PR is merged I can add suitable tests in the test-suite.

FFroehlich commented on docs/src/format.md:
```diff
@@ -36,38 +36,67 @@ A neural network model must consist of two parts to be compatible with the PEtab
+### [Neural Network Input Data](@id hdf5_input_structure)
 
-Potential neural network input files are expected to have the following format:
+Array input data for neural network models is specified in HDF5 format. Each HDF5 file should contain a list of entries, where each entry associates an input Id with datasets. Each dataset consists of the data array and, optionally, the experiment Ids to which it applies. If no experiment Ids are provided, the dataset will be applied to all experiments. Multiple datasets may not be assigned to th ...
```
There should probably be a statement that if experiment ids are specified there needs to be an assignment for each experiment id.

Wouldn't it be more consistent to do this at the level of conditions – I do see pros and cons for either option.

dilpath replied:
> Wouldn't it be more consistent to do this at the level of conditions – I do see pros and cons for either option.

Yes you're right, since we say that NN outputs are specified on the level of conditions, then it makes sense to define their inputs on the level of conditions too. Will switch to conditions then.

> There should probably be a statement that if ~~experiment~~ condition ids are specified there needs to be an assignment for each ~~experiment~~ condition id.

Do you mean each condition ID where the neural network output is relevant, or for all condition IDs?
I think we currently support assigning neural network outputs for only a subset of conditions, so I think we only need to require neural network inputs for that subset of conditions. But yes, I will add this clarification explicitly.

sebapersson replied:
> I think we currently support assigning neural network outputs for only a subset of conditions, so I think we only need to require neural network inputs for that subset of conditions. 

This is correct, and another thing I have to add tests for.

dilpath commented:
Summary of changes since last review

- **added `perm` field in HDF5 files, to specify whether arrays are row-major or column-major**
- frozen layers are now permitted
- dropped input transformations
- dropped arbitrary ML model support
  - rewrote all mentions of “neural network”/”neural net”/”net” to consistently be “NN” in format.md
  - `MLModelStandard` -> `NNModelStandard`
- switched to condition Ids instead of experiment Ids, in the input data HDF5 format
- clarified expectation: input data is required to be specified somewhere in the PEtab SciML problem, for all conditions that involve neural network output

sebapersson commented on docs/src/format.md:
```diff
@@ -1,141 +1,179 @@
 
 !!! tip "For users: Define models in PyTorch"
-    The recommended approach to create a YAML network file is to first define a PyTorch model (`torch.nn.Module`) and use the Python `petab_sciml` library to export this to the YAML format. See the tutorials for examples of this.
+    The recommended approach to create a NN model YAML file is to first define a PyTorch model (`torch.nn.Module`) and use the Python `petab_sciml` library to export this to YAML. See the tutorials for examples of this.
```
Good tip!

sebapersson commented on docs/src/format.md:
```diff
@@ -1,141 +1,179 @@
-- `net1.parameters`
-- `net1.parameters[conv1]`
-- `net1.parameters[conv1].weight`
+- `nn1.parameters`
```
Why nn1 and not nn?

dilpath replied:
Just that, so far in this document, we append numbers to terms to make it clear that the term is an example PEtab identifier, e.g. `conditionId1`, `inputId1`, `layerId1`, `conv1`, so I used `nn1` here too. I think it's nice like this but also fine for me to change this to `nn` if you prefer, since the format of this nested identifier is immediately above it.

sebapersson replied:
Fair, then continue to use `nn1`

sebapersson reviewed (approved):
Thanks, a minor comment. When it is addressed, feel free to merge.
